go.mod

```diff
@@ -6,8 +6,9 @@ toolchain go1.22.3
 
 replace (
 	github.com/containerd/containerd => github.com/containerd/containerd v1.6.27 // for compatibilty with docker 20.10.x
+	github.com/crewjam/saml => github.com/rancher/saml v0.4.14-rancher3
```

Comment:
What is our long term plan regarding this fork, please? Was this validated with the Architecture team? Just asking to understand our goals and if we plan to eventually go back to the original upstream version or not.

I see in our fork that we added some fixes and open (not merged) upstream PRs. Are we working with upstream to get those merged?

My concern is that this is a very sensitive library in terms of security, and it already had 5 CVEs - https://github.com/crewjam/saml/security . By going with the fork we might lose track of future CVE fixes.

Reply:
> Are we working with upstream to get those merged?

The PRs for the fixes we use look to have been ignored for more than a year, both.
And the project has lots of other open PRs which are not attended to.
It feels as if the entire project has fallen out of maintenance.
As such I am unsure how we could work with upstream to get these merged.

> What is our long term plan regarding this fork, please?

Personally I would push for us to search for a better maintained replacement.

> Was this validated with the Architecture team?

Who should I talk to ?

Reply:
@mattfarina WDYT about the situation described by Andreas, please?

Reply:
@andreas-kupries I have a couple questions:

1. Do we have a documented list of changes we are carrying and the reason for each of them? I ask because 1) I am looking to understand the situation and 2) I've run into problems with other forks where we had lost that information and no longer knew.
2. Do we know of any alternative saml packages that are well maintained and stable? I suspect the answer is "no" because we have not looked but I wanted to ask. Just in case.

Reply:
There is currently no documented list of changes, except through the commit messages in the fork.
I am ok with writing such a list. Where should it go ? (Just) in the forked repo, or elsewhere too ?

I know of https://pkg.go.dev/github.com/russellhaering/gosaml2 because during the search
for the OKTA/KC issues it was used in part. The parts were removed again when the first crewjam patch was added,  fixing the mishandling of compressed data.

Some other packages I saw were explicitly marked as experimental, incomplete, etc.

Reply:
Here is a proposed list of changes, to be added to the README.md of the `rancher/saml` fork of `crewjam/saml`

---
# FORK information

This fork of the https://github.com/crewjam/saml repository is rooted in commit
a32b643a25a46182499b1278293e265150056d89 of the origin repository.

This is the head of the `main` branch at the time of writing this document, containing version
`v0.4.14` of the package, i.e. commit b07b16cf83c4171d16da4d85608cb827f183cd79, plus a number of
version bumps for dependent modules, via dependabot.

Compared to the root the fork contains the following changes

  1. Application of pull request https://github.com/crewjam/saml/pull/491

     This fixes the mishandling of compressed data. The XML parser always operated on the input data
     even if it was compressed. While decompression was performed the resulting data was ignored and
     not used.

  1. Application of pull request https://github.com/crewjam/saml/pull/449

     This adds the missing verification of detached signatures to the handling of logout responses.

  1. Beyond the above the work of item 2 was modified/extended to allow an embedded signature to be
     missing when a good, i.e. verified detached signature is found.

     This means that the code now accepts the case of detached signature without embedded signature
     as valid. This is what Keycloak provides.

  1. Going further the work of item 2 was used as template to add the generation of detached
     signatures to logout requests generated by the package.

     Such a missing detached signature was the cause of OKTA not accepting our logout requests. The
     embedded signature had no meaning.

Reply:
And added to the fork, with commit https://github.com/rancher/saml/commit/51375381b65c352cfe0e7635aefb20c27dc5be80

Reply:
So just for my own understanding, https://pkg.go.dev/github.com/russellhaering/gosaml2 doesn't contain all the code and fixes that we need without having the need to still fork it too, right?

Note that our fork https://github.com/rancher/saml doesn't have `main` as the default branch, instead if points to `v0.0.1-rancher1` which can be confusing. Would be good to update the default branch if possible, please.

Reply:
> https://pkg.go.dev/github.com/russellhaering/gosaml2

This package came up during work here as a possible alternative to `crewjam/saml`, which I did not investigate deeply enough to know if it provides everything we need, or not. I saw a fork of `crewjam/saml` as the less risky path than having to replace it with something completely new.

> Would be good to update the default branch if possible, please.

Checking, I find that I do not have the admin permissions necessary to change the default path. Who should I talk to get the change done ?

Reply:
Thanks for the explanation, Andreas.

> I saw a fork of crewjam/saml as the less risky path than having to replace it with something completely new.

Putting this into perspective, it does make sense.

The security implications, ownership and maintainability of such forks, that and others, will eventually be addressed in the Engineering Handbook (CC @mattfarina).

> Checking, I find that I do not have the admin permissions necessary to change the default path. Who should I talk to get the change done ?

You must open a GH issue for EIO.

Reply:
Ok.

Reply:
Done: https://github.com/rancherlabs/eio/issues/2429

Reply:
EIO is done - https://github.com/rancherlabs/eio/issues/2429#issuecomment-2209334653